filter/series_by_tag.go

```diff
@@ -114,42 +116,95 @@ func createMatcher(spec TagSpec) func(string, map[string]string) bool {
 	}
 }
 
+type MatchingFunc func(string, map[string]string) bool
+
 // SeriesByTagPatternIndex helps to index the seriesByTag patterns and allows to match them by metric
 type SeriesByTagPatternIndex struct {
-	patternToMatcher map[string]func(string, map[string]string) bool
+	namesPatternIndex      *PatternIndex
+	nameToPatternMatchers  map[string]map[string]MatchingFunc
+	specialPatternMatchers map[string]MatchingFunc
 }
 
 // NewSeriesByTagPatternIndex creates new SeriesByTagPatternIndex using seriesByTag patterns and parsed specs comes from ParseSeriesByTag
-func NewSeriesByTagPatternIndex(tagSpecsByPattern map[string][]TagSpec) *SeriesByTagPatternIndex {
-	patternToMatcher := make(map[string]func(string, map[string]string) bool)
+func NewSeriesByTagPatternIndex(logger moira.Logger, tagSpecsByPattern map[string][]TagSpec) *SeriesByTagPatternIndex {
+	names := make(map[string]bool)
+	nameToPatternMatchers := make(map[string]map[string]MatchingFunc)
+	specialPatternsToMatcher := make(map[string]MatchingFunc)
+
 	for pattern, tagSpecs := range tagSpecsByPattern {
-		matchers := make([]func(string, map[string]string) bool, 0)
+		matchers := make([]MatchingFunc, 0)
+		var nameTagValue string
+		var hasNameTagWithEqualOperationValue bool
+
 		for _, tagSpec := range tagSpecs {
-			matchers = append(matchers, createMatcher(tagSpec))
+			if tagSpec.Name == "name" && tagSpec.Operator == EqualOperator {
+				nameTagValue = tagSpec.Value
+				names[nameTagValue] = true
+				hasNameTagWithEqualOperationValue = true
+			} else {
+				matchers = append(matchers, createMatcher(tagSpec))
+			}
 		}
 
-		patternToMatcher[pattern] = func(metric string, labels map[string]string) bool {
+		matchingFunc := func(metric string, labels map[string]string) bool {
 			for _, matcher := range matchers {
 				if !matcher(metric, labels) {
 					return false
 				}
 			}
 			return true
 		}
+
+		if hasNameTagWithEqualOperationValue {
+			if _, ok := nameToPatternMatchers[nameTagValue]; ok {
+				nameToPatternMatchers[nameTagValue][pattern] = matchingFunc
```

Comment:
There is only one unique way to build matchingFunc from the pattern. As result, it is not important how many copy of patterns will be, because all of them will have the same matchingFunc.